orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -194,43 +209,133 @@ Status FromTensorProtos(
+
+    if (tensor_proto.data_location() == ONNX_NAMESPACE::TensorProto_DataLocation_EXTERNAL) {
+      std::unique_ptr<ExternalDataInfo> external_data_info;
+      ORT_RETURN_IF_ERROR(ExternalDataInfo::Create(tensor_proto.external_data(), external_data_info));
+      fbs_buffer_size += external_data_info->GetLength();
+    }
+
+    if (bytes > onnxruntime::fbs::utils::kMinimumSizeForExternalData) {  // assuming strings aren't trainable
+      fbs_potential_external_buffer_size += bytes;
+    }
   }
+
   for (const auto& tensor_proto : non_trainable_tensor_protos) {
-    fbs_buffer_size += tensor_proto.ByteSizeLong();
+    auto bytes = tensor_proto.ByteSizeLong();
+    fbs_buffer_size += bytes;
+
+    if (tensor_proto.data_location() == ONNX_NAMESPACE::TensorProto_DataLocation_EXTERNAL) {
+      ORT_RETURN_IF_NOT(tensor_proto.external_data()[2].key() == "length",
+                        "Invalid external data for ", tensor_proto.name());
+      // external_data field is a dictionary of strings
+      // length is stored as a string
+      fbs_buffer_size += std::stoull(tensor_proto.external_data()[2].value());
+    }
+
+    if (bytes > onnxruntime::fbs::utils::kMinimumSizeForExternalData &&
+        tensor_proto.data_type() != ONNX_NAMESPACE::TensorProto_DataType_STRING) {
+      fbs_potential_external_buffer_size += bytes;
+    }
   }
 
   // Align buffer size to 1MB.
+  // TODO: This should probably also have an allowance for the other parts of the tensor if we're trying to avoid
+  // all reallocs during serialization. e.g. name, shape, data type, offset if external data is used.
   fbs_buffer_size = std::max(fbs_buffer_size, m_bytes);
   fbs_buffer_size = ((fbs_buffer_size + m_bytes - 1) / m_bytes) * m_bytes;
+
+  // 2GB is the real limit, but fbs_buffer_size doesn't include space for initializer names and shapes.
+  // for now, arbitrarily use ~1.8GB as a conservative 'safe' value with a 10% buffer.
+  const bool use_external_data = fbs_buffer_size >= external_data_threshold * m_bytes;
+
+  fbs::utils::ExternalDataWriter external_data_writer = nullptr;
+  std::optional<std::ofstream> external_data_stream;
+
+  if (use_external_data) {
+    // 2GB. Have to use external file for data.
+    fbs_buffer_size -= fbs_potential_external_buffer_size;  // reduce fbs buffer size to account for external data
+
+    auto data_path = ExternalCheckpointDataPath(checkpoint_path);
+    external_data_stream = std::ofstream(data_path, std::ios::binary);
+
+    ORT_RETURN_IF(external_data_stream->fail(), "Failed to create checkpoint's external data file: ",
+                  ToUTF8String(data_path));
+
+    // setup the data writer to write aligned data to external_data_stream
+    external_data_writer = [&external_data_stream](int32_t data_type, gsl::span<const uint8_t> bytes,
+                                                   uint64_t& offset) {
+      // for now align everything to 4 or 8 bytes. we can optimize this later if needed.
```

Comment:
Split this out into a helper so it can be used in FromCheckpointState as well

```
static Status WriteExternalData(std::ofstream& external_data_stream, int32_t data_type, gsl::span<const uint8_t> bytes,
                                uint64_t& offset) {
  // for now align everything to 4 or 8 bytes. we can optimize this later if needed.
  int32_t alignment = 4;

  // TODO: Add more special-cased 8 byte types if needed.
  if (data_type == ONNX_NAMESPACE::TensorProto_DataType_INT64 ||
      data_type == ONNX_NAMESPACE::TensorProto_DataType_DOUBLE) {
    alignment = 8;
  }

  int64_t pos = external_data_stream.tellp();

  if (pos % alignment != 0) {
    // 8 bytes of 0's so we can pad to alignment 8 in a single `write`
    constexpr static const uint64_t zeros = 0;
    int64_t padding = alignment - (pos % alignment);
    // skipping validation of this write. doesn't matter if this or the 'real' write below fails. if this does the
    // other will as well as nothing will clear the failure bit in the ofstream in between the calls.
    external_data_stream.write(reinterpret_cast<const char*>(&zeros), padding);
    pos += padding;
  }

  external_data_stream.write(reinterpret_cast<const char*>(bytes.data()), bytes.size());
  // TODO: include error code/message for why it failed. requires platform specific code
  ORT_RETURN_IF(external_data_stream.fail(), "Failed writing external checkpoint data.");


  offset = pos;
  return Status::OK();
};
```
...
```
    external_data_writer = [&external_data_stream](int32_t data_type, gsl::span<const uint8_t> bytes,
                                                   uint64_t& offset) {
      return WriteExternalData(*external_data_stream, data_type, bytes, offset);
    };
```